### DISCUSSION TITLE
TTS - Mixing datasets for FastPitch + HiFiGAN

### DISCUSSION THREAD

**Discussion by gedefet** [NVIDIA-NeMo/NeMo#3678] [Q&A] [answered]
Hi All.

At the end of the FastPitch_Finetuning.ipynb tutorial:

https://github.com/NVIDIA/NeMo/blob/main/tutorials/tts/FastPitch_Finetuning.ipynb

it suggest two different ways to improve performance, one of which is:

_Mix new speaker data with old speaker data
We recommend to train fastpitch using both old speaker data (LJSpeech in this notebook) and the new speaker data. In this case, please modify the .json when finetuning fastpitch to include speaker information:_

but it isn't 100% clear how to do that.

I found another FastPitch_Finetuning.ipynb here: 

https://colab.research.google.com/github/NVIDIA/NeMo/blob/stable/tutorials/tts/FastPitch_Finetuning.ipynb

Which is different than the other. 
It seems that it has two functions defined to do the mixing, but it seems that this colab is abandoned. Nevertheless, I think that has very useful code, and that has tackled this part of mixing datasets.

Is this true? Or do I need something more that is not there?Why it is abandoned?

Thanks!

**Comment by titu1994** [comment]
@Oktai15 any input ?

**Comment by okuchaiev** [comment]
@redoctopus can you please comment on this?

**Comment by redoctopus** [comment, marked as answer]
I've updated the FastPitch fine-tuning notebook in #3954, hopefully the procedure for mixing datasets should be more clear.

As mentioned in the tutorial, we don't explicitly have a code section where you can run this mixing in the notebook itself due to resource limitations on Colab. But if you have a local machine to use, you should be able generate a new manifest and re-run the fine-tuning commands with the few command line changes specified without too much trouble (I hope!).

**Comment by arinaruck** [comment]
Hello!
Is there any info on how to select the amount of "old speaker" data?
Should the totalling duration be the same or should the amount of the old data me scaled proportionally to the amount of the new data?
I have almost 3 hours of the new speaker, so I am not able to provide 60 times that of the old speaker data
And does the data mixing fine-tuning setup scale for the larger amounts of data or is it beneficial only when very limited data of the new speaker is accessible?
@redoctopus, maybe you can comment on this?
Thanks

**Comment by redoctopus** [comment]
Yes, please see the finetuning tutorial mentioned in the original question, specifically the last section "Adding more data": https://github.com/NVIDIA/NeMo/blob/main/tutorials/tts/FastPitch_Finetuning.ipynb

> 5 hours of data from the old speaker should be sufficient for training; it's up to you how much data from the old speaker to use in validation.

We usually do a 1:1 ratio of old to new entries in the manifest, repeating the new entries as necessary to keep the dataset balanced, until you have the total number of samples that you want to train with. (See the tutorial for an example.)

I don't think we've done any formal experiments on the effect of data mixing and scaling the dataset, but the mixing will definitely benefit smaller new speaker datasets more dramatically than for larger datasets. 15 minutes to about an hour of new speaker audio, the mixing helps a lot; after that it becomes less noticeable. And of course it also depends on whether you care about keeping quality of the original speaker's generated speech.

**Reply by gedefet** [reply]
@redoctopus , why do you match the number of samples instead of the total duration of the datasets? I had better results matching duration instead of # of samples. I have even better results decreasing the total old data wrt the new speaker, even with little new speaker data (~25min)

**Reply by redoctopus** [reply]
Duration is certainly more precise! Interleaving them for an equal number of samples was just simpler to write.

> I have even better results decreasing the total old data wrt the new speaker, even with little new speaker data (~25min)

This is interesting, thanks for bringing it up. We haven't done much in the way of experimentation for this--mostly we were comparing "no mixing (new speaker data only)" vs. "1:1 mixing (new and old speaker data)" and saw that the mixing was noticeably better for most of our new speakers (i.e. 15m, 30m, 60m of non-professionally recorded audio, varying sound quality). As we got to 60m or so of new speaker data, the difference was much less, at least to my non-professional ear. It would be interesting to see (hear?) a detailed analysis of mixing ratios and resulting speaker quality. I wonder if the recording quality of the new speaker has some effect on the benefit, too...

**Reply by gedefet** [reply]
@redoctopus in my case, I'm using spotify podcasts recorded in a studio, and then processed with an audio tool. But there is certainly a distance to a "professional" recording...

I also use a smaller subset for HiFiGAN,  reaching a point with good quality in the sense of audio quality

**Comment by helgadnestr** [comment]
@gedefet could you tell me please about your experience - do you have success finetuning FastPitch+Hifigan with a small new dataset duration 15-30 minutes? I tryed to fine-tuning one FastPitch variation but the quality is not good enough and I suspect that fastpitch only gives good quality when training from scratch with large datasets. I would be grateful for any information about your experience.

**Reply by gedefet** [reply]
@helgadnestr you have to have a VERY good dataset. It has to be nearly perfect. That is the key

**Reply by helgadnestr** [reply]
@gedefet Did you manage to get such the perfect dataset? Could you send me some examples of records from the dataset and generated records trained on it on my email? and how duration of this new dataset did you have when was training?
If I understand correctly, this is a drawback FastPitch - it needs perfect records in the dataset..? In one project (coqui) two networks FastPitch and Vits were trained on the dataset VCTK (each speaker recorded about 400 examples and NOT perfect quality), generated records of Vits were very well, but FastPitch generated records were bad, they had a lot of metal and synthetics.

**Reply by gedefet** [reply]
The quality should be like LJSpeech or HiFiTTS dataset. I recorded it on a professional studio

**Reply by helgadnestr** [reply]
@gedefet  thank you, I understood about quality dataset.. but may be you can sent me examples generated records, it's important for me to know that it sound like human after fine-tuning with small good dataset.

**Reply by gedefet** [reply]
Unfortunately I can't, as they are private, non-public samples

**Reply by helgadnestr** [reply]
No problem, maybe someone else can answer me.. thanks for the answers!
